tink worker command line args

displague opened the pull request:
## Description

This PR makes most constants and environment variables for tink-worker configurable from the command line, environment variables, and falls back to a `.tinkerbell.yaml` (`worker-id: 1234`) (or .json) file.

```
$ ./tink-worker -w 212
Error: required flag(s) "docker-registry", "registry-password", "registry-username" not set
Usage:
  tink-worker [flags]

Flags:
  -r, --docker-registry string     Sets the Docker registry (DOCKER_REGISTRY)
  -h, --help                       help for tink-worker
  -s, --max-file-size int          Maximum file size in bytes (MAX_FILE_SIZE) (default 10485760)
  -m, --max-retry int              Maximum number of retries to attempt (MAX_RETRY) (default 3)
  -p, --registry-password string   Sets the registry-password (REGISTRY_PASSWORD)
  -u, --registry-username string   Sets the registry username (REGISTRY_USERNAME)
  -i, --retry-interval duration    Retry interval in seconds (RETRY_INTERVAL) (default 3ns)
  -v, --version                    version for tink-worker
  -w, --worker-id string           Sets the worker id (WORKER_ID)
  -l, --worker-log-level string    Sets the worker log level (panic, fatal, error, warn, info, debug, trace) (WORKER_LOG_LEVEL) (default "info")
  -t, --worker-timeout duration    Max duration to wait for worker to complete (WORKER_TIMEOUT)
```

Notably, `TINKERBELL_CERT_URL` lookups are checked from a client that is defined in a parent package. This variable has not been made configurable in the same way.

This is similar in spirit to #266 

Considerable refactoring was done to clean up the use of global variables. The `internal` pkg contains a `Worker` and `RegistryConnDetails` struct which are context bags for the methods that needed these global variables. I imagine we will continue to evolve these types to find a better shape.



## Why is this needed

When run from various environments, users may have an easier time manipulating either configuration files, command line arguments, or the environment. Also, by defining these variables with cobra/viper, users can leverage `--help` and we can eventually use this to generate man pages, shell auto-completions, and markdown docs.

Fixes: #

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->


## How are existing users impacted? What migration steps/scripts do we need?

<!--- Fixes a bug, unblocks installation, removes a component of the stack etc -->
<!--- Requires a DB migration script, etc. -->

## Known Bugs

* `--help` does not include the environment variable hints which are shown when invalid args are supplied 🤔  (these hints are generated at runtime, we could just move them into the text description)

## Checklist:

I have:

- [ ] updated the documentation and/or roadmap (if required)
- [ ] added unit or e2e tests
- [ ] provided instructions on how to upgrade

gianarb commented:
That's a great effort @displague !! Now we have a page in our documentation for each service. You analyzed the topic well and you know the various ways to configure the tink-worker, do you mind adding a section in the tink-worker doc about this? https://tinkerbell.org/docs/services/tink/ Thanks

displague commented:
@gianarb That sounds like a good idea. I'll need to become a little more familiar with tink-worker before I can do that. I need to add tests to this pull request (and actually test it out myself 😅 ) before doing so.

detiber commented on cmd/tink-worker/internal/registry.go:
```diff
@@ -0,0 +1,75 @@
+	if r.registry == "" {
+		return nil, errors.New("required DOCKER_REGISTRY")
+	}
+	c, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
```
This will likely need to be updated to work with non-env based registry url, I think potentially just adding client.WithHost(r.registry) would be sufficient

displague replied:
That sounds about right and good.

Is `client.WithAPIVersionNegotiation` a default option? I'll look into it.

This `client.FromEnv` options resembles the Tink NewClient example I offered up in feedback elsewhere, it seems I was inspired.

displague replied:
I'm going with `c, err := client.NewClientWithOpts(client.FromEnv, client.WithHost(r.registry), client.WithAPIVersionNegotiation())`. I suspect that including `client.FromEnv` may be needed for some other DOCKER_ variables that we are not exposing elsewhere.

```
Supported environment variables:
DOCKER_HOST to set the url to the docker server.
DOCKER_API_VERSION to set the version of the API to reach, leave empty for latest.
DOCKER_CERT_PATH to load the TLS certificates from.
DOCKER_TLS_VERIFY to enable or disable TLS verification, off by default.
```

(Seems we may have invented DOCKER_REGISTRY for what DOCKER_HOST was already providing?)

detiber replied:
There is going to be a slight distinction between DOCKER_HOST and our use of DOCKER_REGISTRY, in this case DOCKER_HOST is referring to the docker daemon, which we would likely want to generally *not* override here in favor of the local docker daemon

mmlb commented on cmd/tink-worker/internal/registry.go:
```diff
@@ -0,0 +1,75 @@
+
+// pullImage outputs to stdout the contents of the requested image (relative to the registry)
+func (w *RegistryConnDetails) pullImage(ctx context.Context, cli *client.Client, image string) error {
+	registry := w.registry
```
we should either use `registry` in line 58 or `w.registry` on line 66.

displague replied:
I think this was some attempt to keep my initial diff noise down between local refactors. What a mess.

mmlb commented on cmd/tink-worker/internal/worker.go:
```diff
@@ -1,27 +1,30 @@
-package main
+package internal // import "github.com/tinkerbell/tink/cmd/tink-worker/internal"
```
this is the first import statement in this repo, we should discuss this separately and/or have it in every file right?

displague replied:
fair enough, I'll remove the `import` comment. I had some early problems getting this internal package to work (ultimately unrelated) and I became very explicit in the process.

displague replied:
https://golang.org/pkg/cmd/go/#hdr-Import_path_checking

mmlb commented on cmd/tink-worker/internal/worker.go:
```diff
@@ -1,27 +1,30 @@
 	"github.com/sirupsen/logrus"
-	pb "github.com/tinkerbell/tink/protos/workflow"
 	"google.golang.org/grpc/status"
+
```
another line should go away

mmlb commented on cmd/tink-worker/main.go:
```diff
@@ -2,93 +2,22 @@ package main
-	retryIntervalDefault = 3
-	retryCountDefault    = 3
-)
 
```
remove

displague commented on cmd/tink-cli/cmd/root.go:
```diff
@@ -37,6 +38,10 @@ func Execute(version string) error {
 // initConfig reads in config file and ENV variables if set.
 func initConfig() {
 	viper.AutomaticEnv() // read in environment variables that match
+	viper.SetConfigName(".tinkerbell")
+	viper.AddConfigPath("/etc/tinkerbell")
+	viper.AddConfigPath("$HOME/.config/tinkerbell")
+	viper.AddConfigPath(".")
```
In Go, iteration over map indexes is unpredictable, these paths are appended to a slice of strings - order is therefore fixed.

That said, these paths are in reverse order for no good reason (I copied this snippet out of the viper example documentation and modified the paths, I assumed it was in the logical precedence order).

AddConfigPath puts the latest entry at the end of the list, `ReadInConfig` uses the first path in the list that has a matching file (no merging of subsequent matches).

I'll update this accordingly.  Thanks for spotting this!

displague commented on cmd/tink-cli/cmd/root.go:
```diff
@@ -37,6 +38,10 @@ func Execute(version string) error {
 // initConfig reads in config file and ENV variables if set.
 func initConfig() {
 	viper.AutomaticEnv() // read in environment variables that match
+	viper.SetConfigName(".tinkerbell")
```
`tinkerbell-worker` it is.

displague replied:
`tink-worker`, that is.

displague commented on cmd/tink-worker/cmd/root.go:
```diff
@@ -0,0 +1,232 @@
+	"github.com/spf13/pflag"
+	"github.com/spf13/viper"
+	"google.golang.org/grpc"
+
```
👍  I'd argue in favor of `goimports -w -local github.com/tinkerbell/tink`, but that should be dealt with separately.

displague commented on cmd/tink-worker/cmd/root.go:
```diff
@@ -0,0 +1,232 @@
+
+	cmd.PersistentFlags().VisitAll(func(f *pflag.Flag) {
+		// Bind flag names to proper env var name, unhyphenated are handled by AutomaticEnv
+		if strings.Contains(f.Name, "-") {
+			envVar := strings.ToUpper(strings.ReplaceAll(f.Name, "-", "_"))
+			f.Usage = f.Usage + " (" + envVar + ")"
+			err := v.BindEnv(f.Name, envVar)
+			if err != nil {
+				errors = append(errors, err)
+				return
+			}
+		}
```
I would think that `v.AutomaticEnv()` makes this loop entirely unneeded, but it did not work in practice. 
I've suspected that the `-` vs `_` may be behind this, or casing.
I'll give `SetEnvKeyReplacer` a try with `AutomaticEnv` and perhaps we can get rid of this loop and manual binding.

displague replied:
I was also taking advantage of this loop to add the environment variable name to the field description, but that probably makes more sense to do in the original string. (the default values are added to the description dynamically, so this wasn't too zany of an approach).

displague replied:
Have a look at what I've got in here now.

I've been running this as:

```sh
$ echo -e "---\nregistry-username: foo"> tink-worker.yaml
$ TINK_WORKER_ID=123 go run ./... --docker-registry=123
Using config file: /Users/marques/src/tink/cmd/tink-worker/tink-worker.yaml
Error: required flag(s) "registry-password" not set
```

This latest commit is as concise as I could get viper+cobra working together to provide all of the values.

displague commented on cmd/tink-worker/cmd/root.go:
```diff
@@ -0,0 +1,232 @@
+			}
+		}
+
+		// Apply viper config values when cobra has no value and viper does
+		if !f.Changed && v.IsSet(f.Name) {
+			val := v.Get(f.Name)
+			if err := cmd.Flags().Set(f.Name, fmt.Sprintf("%v", val)); err != nil {
+				errors = append(errors, err)
+				return
+			}
+		}
```
The default logic wasn't working for me. https://github.com/spf13/viper/issues/969#issuecomment-687871054

I ended up following this pattern https://carolynvanslyck.com/blog/2020/08/sting-of-the-viper/

displague replied:
If you have a better handle on how to manage this, I'd gladly simplify this to what I expected viper to do out of the box.

displague replied:
Going with what I said in https://github.com/tinkerbell/tink/pull/287#discussion_r489151751 -- if you have better ideas, I'm all ears.

displague commented on cmd/tink-worker/internal/action.go:
```diff
@@ -0,0 +1,116 @@
+		config.Cmd = cmd
+	}
+
+	wfDir := dataDir + string(os.PathSeparator) + wfID
```
I inherited this line from the file rename,  
https://github.com/tinkerbell/tink/pull/287/files#diff-f776dc055922cd9d519a49a712a9d22fL181

Glad to clean it up 👍

displague commented on cmd/tink-worker/cmd/root.go:
```diff
@@ -0,0 +1,232 @@
+
+	rootCmd.PersistentFlags().DurationP("retry-interval", "i", retryIntervalDefault, "Retry interval in seconds")
+
+	rootCmd.PersistentFlags().DurationP("worker-timeout", "t", 0, "Max duration to wait for worker to complete")
+
+	rootCmd.PersistentFlags().IntP("max-retry", "m", retryCountDefault, "Maximum number of retries to attempt")
+
+	rootCmd.PersistentFlags().Int64P("max-file-size", "s", defaultMaxFileSize, "Maximum file size in bytes")
+
+	rootCmd.PersistentFlags().StringP("worker-log-level", "l", "info", "Sets the worker log level (panic, fatal, error, warn, info, debug, trace)")
+
+	rootCmd.PersistentFlags().StringP("worker-id", "w", "", "Sets the worker id")
```
👍  I had a shared tinkerbell config file in mind when I had the prefix here. A `tinkerbell-worker.yaml` won't have the same ambiguity.

displague commented on cmd/tink-worker/cmd/root.go:
```diff
@@ -0,0 +1,232 @@
+func initializeLogger(logger *logrus.Logger, level string) {
+	if level != "" {
+		switch strings.ToLower(level) {
+		case "panic":
```
Apparently, there is a `ParseLevel` function that mirrors this switch statement.
```go
		if lvl, err := logrus.ParseLevel(level); err == nil {
			logger.SetLevel(lvl)
			return
		}
```

displague replied:
The levels themselves are iota ints, so I'm not sure we could `.VarP()` them.

displague replied:
Closing this thread since LOG_LEVEL (and the associated changes) went away with the rebase on #271.

displague commented:
@mmlb @detiber I've addressed most of your concerns and commented on the ones that I didn't enact your recommendations.

Do either of you have an environment where you can connect this worker to the service and interact with it as you normally would?

detiber commented on cmd/tink-worker/internal/worker.go:
```diff
@@ -220,7 +357,7 @@ func processWorkflowActions(client pb.WorkflowSvcClient) error {
 		}
 		// sleep for 3 seconds before asking for new workflows
-		time.Sleep(retryInterval * time.Second)
+		time.Sleep(w.retryInterval * time.Second)
```
```suggestion
		<-time.After(w.retryInterval * time.Second)
```

displague commented on cmd/tink-worker/cmd/root.go:
```diff
@@ -0,0 +1,171 @@
+	rootCmd.PersistentFlags().StringP("id", "i", "", "Sets the worker id")
+	must(rootCmd.MarkPersistentFlagRequired("id"))
+
+	rootCmd.PersistentFlags().StringP("docker-registry", "r", "", "Sets the Docker registry")
+	must(rootCmd.MarkPersistentFlagRequired("docker-registry"))
+
+	rootCmd.PersistentFlags().StringP("registry-username", "u", "", "Sets the registry username")
+	must(rootCmd.MarkPersistentFlagRequired("registry-username"))
+
+	rootCmd.PersistentFlags().StringP("registry-password", "p", "", "Sets the registry-password")
+	must(rootCmd.MarkPersistentFlagRequired("registry-password"))
```
I agree that too many of the parameters are required now, and that anonymous registry access would be beneficial.

In the base of this PR, the current flow requires these variables:
main -> processWorkflowActions -> initializeDockerClient -> requires DOCKER_REGISTRY
main -> processWorkflowActions -> executeAction -> pullActionImage -> requires REGISTRY_USERNAME, REGISTRY_PASSWORD

I'd rather leave making them option a future task, especially if it requires changes elsewhere.

displague replied:
In this PR, I've brought those requirements closer to the surface.

displague commented on cmd/tink-worker/cmd/root.go:
```diff
@@ -0,0 +1,171 @@
+	v.SetConfigName("tink-worker")
+	v.AddConfigPath("/etc/tinkerbell")
+	v.AddConfigPath(".")
+	v.SetEnvPrefix("TINK_WORKER")
```
mm, yes. I think I would argue in favor of `RETRY_INTERVAL`, `MAX_RETRY`, `DOCKER_REGISTRY`, `REGISTRY_USERNAME`, `REGISTRY_PASSWORD` having a `TINK_WORKER` prefix.

If I remove the prefix, `TINK_WORKER_ID` would become `ID` which is too ambiguous for env usage.

What does @tinkerbell/site-eng think?

displague replied:
@mmlb ?

mmlb replied:
I like having external things not have a prefix so that a common env can be setup and by used by all services (RETRY_INTERVAL, MAX_RETRY mostly). Kind of like being able to apply things globally or as wide as possible. Helps script things up too. Not strongly held though.

displague replied:
It's certainly easier for this PR to not introduce breaking changes on the environment variables.

While I don't prefer the ambiguous names, I do prefer not breaking the API.

I'll update this to not include the `TINK_WORKER` prefix.

displague commented on cmd/tink-worker/cmd/root.go:
```diff
@@ -0,0 +1,171 @@
+			user, _ := cmd.PersistentFlags().GetString("registry-username")
+			pwd, _ := cmd.PersistentFlags().GetString("registry-password")
+			registry, _ := cmd.PersistentFlags().GetString("docker-registry")
+
```
`TINKERBELL_CERT_URL` is defined in `/client/main.go` above what this change `tink-worker` change set out to manipulate. I am interested in seeing the client become configurable (without global environment), but I didn't want to introduce that change here.

displague commented on cmd/tink-worker/internal/action.go:
```diff
@@ -0,0 +1,119 @@
+func (w *Worker) createContainer(ctx context.Context, cmd []string, wfID string, action *pb.WorkflowAction) (string, error) {
+	registry := w.registry
+	config := &container.Config{
+		Image:        registry + "/" + action.GetImage(),
```
Noted in https://github.com/tinkerbell/tink/issues/303

displague commented on cmd/tink-worker/internal/action.go:
```diff
@@ -0,0 +1,119 @@
+func (w *Worker) createContainer(ctx context.Context, cmd []string, wfID string, action *pb.WorkflowAction) (string, error) {
+	registry := w.registry
+	config := &container.Config{
+		Image:        registry + "/" + action.GetImage(),
```
I double-checked that this will always use forward slashes. 👍

displague commented on cmd/tink-worker/internal/action.go:
```diff
@@ -0,0 +1,119 @@
+		Tty:          true,
+		Env:          action.GetEnvironment(),
+	}
+	if cmd != nil {
+		config.Cmd = cmd
+	}
```
Nice improvement over https://github.com/tinkerbell/tink/pull/287/files#diff-f776dc055922cd9d519a49a712a9d22fL185-L187, I'll add it.

displague commented on cmd/tink-worker/internal/action.go:
```diff
@@ -0,0 +1,119 @@
+
+func startContainer(ctx context.Context, l log.Logger, cli *client.Client, id string) error {
+	l.With("containerID", id).Debug("starting container")
+	err := cli.ContainerStart(ctx, id, types.ContainerStartOptions{})
+	if err != nil {
```
What are your thoughts on `return errors.Wrap(cli.ContainerStart(ctx, id, types.ContainerStartOptions{}, "DOCKER START")` instead? (I know some people prefer this, but I never really liked doing "work" in the `Wrap` call. I prefer the `err :=` being on a separate line, but I'll go with the one-liner approach unless I get feedback against.

displague commented on cmd/tink-worker/internal/action.go:
```diff
@@ -0,0 +1,119 @@
+	case err := <-errC:
+		l.Error(err)
+		failedActionStatus <- pb.ActionState_ACTION_FAILED
+	}
```
That sounds reasonable to me. A "*wait*" function does seem to be the best place to test the context.

I suspected that `errC` and `ContainerWait` may have been more involved in the `ctx`, but it just seems to pass it along.

I'll use ACTION_TIMEOUT and emit an error log in that case.  I'll generically state that the "context was closed".

displague replied:
Ended up doing something closer to the previous `switch`, `l.Error(ctx.Error())`

displague commented on cmd/tink-worker/internal/worker.go:
```diff
@@ -45,28 +48,162 @@ type WorkflowMetadata struct {
+	if action.Timeout > 0 {
+		timeCtx, cancel = context.WithTimeout(ctx, time.Duration(action.Timeout)*time.Second)
+	} else {
+		timeCtx, cancel = context.WithTimeout(ctx, 1*time.Hour)
```
With this PR, we have a default timeout being applied in cmd/root.go, this ctx is being used throughout the end-to-end worker process in this PR.

`ctx` here is already bound to a timer set from the `timeout` worker argument (which has a 1 hour default, unless that timeout is overridden to "0" with the changes in this PR).

What you've pointed out here is that I overlooked that aspect of the refactor and allowed a secondary timer to be created here which can not be configured.  `action.Timer` should either be configured with a secondary `action-timeout` parameter or it should just inherit the time remaining on the ctx supplied (perhaps action.Timer can be removed entirely).  

I suspect that the condition here should have allowed for an untimed action, dependent on the ctx timer alone, which is the behavior that cmd/root.go now provides.

What you've pointed our here has made me aware of one or two places where I missed a `context.Background()` that should be using the supplied `ctx` instead.

displague replied:
I see now. Actions have their own timeouts, so these do not need to be expressed as worker arguments. 

If the worker is run with a timeout value, the worker should timeout according to the operator's wishes (worker command-line), not the user wishes (action timeout).

displague replied:
Given that, I think I'll leave this block alone. I don't think that a "0" action timeout should result in a 1 hour default timeout. I would opt for this to me "no action timeout" (defer to the worker timeout, if resent).

But I'll leave that for someone else to solve.

displague commented on cmd/tink-worker/internal/worker.go:
```diff
@@ -45,28 +48,162 @@ type WorkflowMetadata struct {
+		}
+		return status, errors.Wrap(err, "DOCKER_WAIT")
 	}
+	rerr := removeContainer(ctx, l, cli, id)
```
That makes sense.

I'm getting worried about the level of changes that I am making from the original (that this may become too complex to review): https://github.com/tinkerbell/tink/pull/287/files#diff-f776dc055922cd9d519a49a712a9d22fL75

detiber replied:
Would it be possible to break the changes down into more easily reviewable chunks?

displague replied:
It looks like we are burying the original "wait" error on wait failure (without logging it), by replacing it with the removal error when removal fails.

If I don't defer the removal, I could wrap the errors and log both failures.  

I'll give the deer a shot.  Even though we may be buring the removal error, it will be deferred and logged, and the original "wait" error will be surfaced.  This should make the code and error easier to understand.

displague replied:
The `removeContainer` calls at the bottom of the function don't even make it into the returned error (they are only logged).  I think defer is the right call based on the existing usage.

displague replied:
I split up the commits a little, @detiber. 

The first commit now captures the `git mv` to `internal/`, which means that the second commit reflects the `actions.go` changes more succinctly.  I could do something similar to keep the "execute" function defined in the same original position as "executeAction" (likewise for capture logs) if that would help.

The third commit (fixup) represents the changes from your latest round of review and includes linter fixes needed to make `make verify` happy.  

The last commit moves all of the arguments from Persistent to local. I didn't venture down the other ways of getting viper to handle things. I think that can wait for a further cleanup, perhaps in your tink-server viper PR.

mmlb commented on cmd/tink-worker/cmd/root.go:
```diff
@@ -0,0 +1,173 @@
+const (
+	retryIntervalDefault        = 3
+	retryCountDefault           = 3
+	defaultMaxFileSize    int64 = 10485760 //10MB ~= 10485760Bytes
```
even better `10 * 1024 * 1024` and we let Go calculate it for us :D

mmlb commented on cmd/tink-worker/cmd/root.go:
```diff
@@ -0,0 +1,173 @@
+)
+
+const (
+	retryIntervalDefault        = 3
+	retryCountDefault           = 3
```
rename to defaultRetryInterval and defaultRetryCount to match other 2?

mmlb commented on cmd/tink-worker/cmd/root.go:
```diff
@@ -0,0 +1,173 @@
+	v := viper.New()
+	v.AutomaticEnv()
+	v.SetConfigName("tink-worker")
+	v.AddConfigPath("/etc/tinkerbell")
+	v.AddConfigPath(".")
```
did we lose somewhere in $HOME?

displague replied:
I could add `$HOME/.config/something/` back, I thought it might be easier to figure that out later. (the preference for a $HOME subpath and a filename in that path may complicate this).

mmlb replied:
makes sense

mmlb commented on cmd/tink-worker/cmd/root.go:
```diff
@@ -0,0 +1,173 @@
+
+// NewRootCommand creates a new Tink Worker Cobra root command
+func NewRootCommand(version string, logger log.Logger) *cobra.Command {
+	must := func(err error) {
+		if err != nil {
+			logger.Fatal(err)
+		}
+	}
```
move to closer to first use?

mmlb commented on cmd/tink-worker/cmd/root.go:
```diff
@@ -0,0 +1,173 @@
+			return nil, err
+		}
+	} else {
+		fmt.Fprintln(os.Stderr, "Using config file:", v.ConfigFileUsed())
```
Why do we log using a config file, if non was found?

displague replied:
If there was an error reading in the config, it could be that none of the config files exist - this error is ok, and is ignored. If there was any other kind of config loading error then we return that error.

If we successfully loaded a config file, we emit that to stderr.

What are you proposing @mmlb? I could log when no config file was found (though none is needed), and I could log when a config file is used, using the common logging format (this PR started before that was introduced and that may explain why I used Println here).

displague replied:
I'm passing the logger in now and logging all three possibilities.

mmlb replied:
:+1:

mmlb commented on cmd/tink-worker/cmd/root.go:
```diff
@@ -0,0 +1,173 @@
+	})
+
+	if len(errors) > 0 {
+		errs := []string{}
+		for _, err := range errors {
+			errs = append(errs, err.Error())
+		}
+		return fmt.Errorf(strings.Join(errs, ", "))
```
I swear I've seen something like multierr in the stdlib and keep thinking we should use it and then never find it again.

displague replied:
https://github.com/hashicorp/go-multierror is the one I am aware of

displague replied:
https://pkg.go.dev/go.uber.org/multierr is in our go.sum, but not used directly.

gianarb replied:
are you speaking about Wrap/Unwrap @mmlb https://blog.golang.org/go1.13-errors ?

mmlb replied:
@gianarb no, I've used hashicorp multierror and I swear I've thought there one similar in the stdlib. Didn't find one when I went looking, probably just my bad memory.

detiber commented on cmd/tink-worker/internal/registry.go:
```diff
@@ -0,0 +1,71 @@
+}
+
+// NewClient uses the RegistryConnDetails to create a new Docker Client
+func (r *RegistryConnDetails) NewClient() (*client.Client, error) {
+	if r.registry == "" {
+		return nil, errors.New("required DOCKER_REGISTRY")
+	}
+	c, err := client.NewClientWithOpts(client.FromEnv, client.WithHost(r.registry), client.WithAPIVersionNegotiation())
```
```suggestion
func (r *RegistryConnDetails) NewClient() (*client.Client, error) {
	c, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
```

detiber replied:
WithHost here is setting the docker daemon host rather than the registry url.

displague replied:
Removed WithHost in the latest push (which has been rebased).

detiber reviewed (approved):
lgtm, thanks for all the hard work you put into this one @displague!
